### PR TITLE
x86: add UEFI bootable ext4 image automation

### PR DESCRIPTION
**Pull request by alive4ever**
With these changes, the build system should be able output UEFI bootable ext4 images for x86 targets.

### CODE REVIEW

**Comment by nbd168** [comment]
Please get rid of the grub2 code duplication. You could make subdirectories 'generic, efi-amd64, efi-i386' under package/boot/grub2 and put a .mk include file there that has the common code.

Also, you probably don't need separate packages for efi-amd64 and efi-i386, you could just make one that detects the selected architecture.

Regarding the DRM removal, I'm not sure if that's a good idea. LEDE is used on more than just embedded systems.

**Comment by alive4ever** [comment]
> Please get rid of the grub2 code duplication. You could make subdirectories 'generic, efi-amd64, efi-i386' under package/boot/grub2 and put a .mk include file there that has the common code.

Done.

> Also, you probably don't need separate packages for efi-amd64 and efi-i386, you could just make one that detects the selected architecture.

You're right. I can just modify grub2 Makefile to build on a subdirectory.

> Regarding the DRM removal, I'm not sure if that's a good idea. LEDE is used on more than just embedded systems.

Done. I am just enabling ``CONFIG_FB_SIMPLE`` for the generic target to make use of grub's framebuffer.

Let me know if there is more to do.

**Comment by alive4ever** [comment]
Rebased with latest HEAD. Let me know if there is more to fix.

**Comment by sduponch** [comment]
Hi, i have done a similar work here https://github.com/openwrt/openwrt/pull/391 that support squashfs with grub2 dedup.

**Comment by alive4ever** [comment]
My changes here only generates uefi bootable image with ext4 filesystem. I try to minimize 
changes that introduces new tool or package, i.e. instead of creating a new grub2-efi, I just 
modify the grub2 makefile to compile inside a subdirectory.

> I note that your are not converting your MBR partition to a GPT one, is this reliable ?

This makes it possible for EFI image to be bootable on both EFI and legacy system and 
prevent ``sysupgrade`` warning about invalid image.

``sysupgrade`` will check for ``eb48`` or ``eb63`` magic header presence to make sure that 
the image is bios bootable before attempting to flash the image.

For GPT images, a separate pull request is necessary to add ``gdisk`` as build tools, which 
is beyond what this pull request scope.

**Comment by alive4ever** [comment]
> Maybe we can work together...
You may review my changes as the starting point.

**Comment by sduponch** [comment]
Ok, my method append EFI specifics partitions (EF00 and EF01) to the legacy bios image (that is only converted to GPT with the MBR in protective mode using gdisk). So the magic is always present and make the image compatible with both bios and efi system without warning from sysupgrade.

The drawbacks is that method make much bigger images and require some grub trick (search file) to deal with GPT partition signature that is not currently supported by grub.

**Comment by alive4ever** [comment]
Still figuring out how to deal with freetype requirement here. If I can't figure it out, I'll drop freetype and grub-mkfont entirely and let the uefi boot without loading unifont.

**Comment by alive4ever** [comment]
Adding `libfreetype` from `packages` feed fix `freetype2` dependency, although it is unnecessary since freetype is not required on the `grub` runtime. It is just a tool to create `unifont.pf2` during host build so that grub can display boot menu better.

Is this acceptable? Or should I add `freetype2` as host tool to satisfy `freetype2` build dependency?

**Comment by alive4ever** [comment]
As an additional notice, this branch will fail during `gzip` image phase. I'll figure something to fix it.

**Comment by pprindeville** [comment]
Still hoping to see this get merged soon.  @alive4ever What's the progress on fixing the `gzip` phase?

File: package/boot/grub2/Makefile [hunk 1]
```diff
@@ -4,7 +4,6 @@
 # This is free software, licensed under the GNU General Public License v2.
 # See /LICENSE for more information.
 #
-
```
Comment (pprindeville) [inline package/boot/grub2/Makefile]: I'd avoid whitespace only changes.

File: package/boot/grub2/Makefile [hunk 2]
```diff
@@ -20,6 +19,7 @@ PKG_SOURCE_URL:=http://alpha.gnu.org/gnu/grub \
 PKG_HASH:=053bfcbe366733e4f5a1baf4eb15e1efd977225bdd323b78087ce5fa172fc246
```
Comment (pprindeville) [inline package/boot/grub2/Makefile]: Do we need a `PKG_VERSION` bump?

File: package/boot/grub2/Makefile [hunk 3]
```diff
@@ -51,44 +51,174 @@ define Package/grub2-editenv/description
...
+  FILE:=$(GRUB_FONT_SOURCE)
+  URL:=$(GRUB_FONT_SOURCE_URL)
+  HASH:=$(GRUB_FONT_SOURCE_HASH)
+endef
+endif
+
 HOST_BUILD_PREFIX := $(STAGING_DIR_HOST)
 
-CONFIGURE_VARS += \
+HOST_CONFIGURE_CMD := $(BASH) ../configure
```
Comment (pprindeville) [inline package/boot/grub2/Makefile]: Is the `$(BASH) ` strictly necessary?  Shouldn't the `./configure` call out whatever interpreter it needs with `#!`?  If it doesn't, it's broken.

File: package/boot/grub2/Makefile [hunk 4]
```diff
@@ -51,44 +51,174 @@ define Package/grub2-editenv/description
...
+
+BIOS_CONFIGURE_VARS += \
 	grub_build_mkfont_excuse="don't want fonts"
 
-CONFIGURE_ARGS += \
+BIOS_CONFIGURE_ARGS += \
```
Comment (pprindeville) [inline package/boot/grub2/Makefile]: Shouldn't this be `HOST_BIOS_CONFIGURE_ARGS` (like below)?  I'm not seeing `BIOS_CONFIGURE_ARGS` being used anywhere...

File: package/boot/grub2/Makefile [hunk 5]
```diff
@@ -51,44 +51,174 @@ define Package/grub2-editenv/description
...
 
-HOST_CONFIGURE_ARGS += \
+HOST_BIOS_CONFIGURE_ARGS += \
 	--disable-grub-mkfont \
 	--target=$(REAL_GNU_TARGET_NAME) \
+	--prefix="$(STAGING_DIR_HOST)" \
+	--sbindir="$(STAGING_DIR_HOST)/bin" \
+	--disable-werror \
+	--disable-libzfs \
+	--disable-nls
```
Comment (pprindeville) [inline package/boot/grub2/Makefile]: We usually don't do this explicitly, do we?  We have `$(DISABLE_NLS)` for this...

File: package/boot/grub2/Makefile [hunk 6]
```diff
@@ -51,44 +51,174 @@ define Package/grub2-editenv/description
...
+
+ifneq ($(CONFIG_GRUB_EFI),)
+EFI_CONFIGURE_ARGS += \
```
Comment (pprindeville) [inline package/boot/grub2/Makefile]: Where does `$(EFI_CONFIGURE_ARGS)` get used?

File: target/linux/x86/image/gen_image_efi.sh [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+KERNELDIR="$3"
+ROOTFSSIZE="$4"
+ROOTFSIMAGE="$5"
+ALIGN="$6"
+
+rm -f "$OUTPUT"
+
+head=16
+sect=63
+cyl=$(( ($KERNELSIZE + $ROOTFSSIZE) * 1024 * 1024 / ($head * $sect * 512)))
```
Comment (pprindeville) [inline target/linux/x86/image/gen_image_efi.sh]: For consistency, close with `) ))`.

File: target/linux/x86/image/gen_image_efi.sh [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+ROOTFSOFFSET="$(($3 / 512))"
+ROOTFSSIZE="$(($4 / 512))"
+
+[ -n "$PADDING" ] && dd if=/dev/zero of="$OUTPUT" bs=512 seek="$ROOTFSOFFSET" conv=notrunc count="$ROOTFSSIZE"
+dd if="$ROOTFSIMAGE" of="$OUTPUT" bs=512 seek="$ROOTFSOFFSET" conv=notrunc
+
+[ -n "$NOGRUB" ] && exit 0
+
+echo "Creating FAT formatted boot image partition"
+dd if=/dev/zero bs=512 count=0 seek="$(($KERNELSIZE/512))" of="$OUTPUT.kernel" &&
```
Comment (pprindeville) [inline target/linux/x86/image/gen_image_efi.sh]: Please add `\` onto continuing lines, and indent `&&` continued lines.

File: target/linux/x86/image/gen_image_efi.sh [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+mkfs.fat -n BOOT "$OUTPUT.kernel" &&
+echo "FAT formatted boot image partition created"
+echo "Copying boot and EFI directory into boot image partition"
+mcopy -s -i "$OUTPUT.kernel" "$KERNELDIR/boot" "$KERNELDIR/EFI" ::/ || echo "Copying boot and EFI directory failed!"
```
Comment (pprindeville) [inline target/linux/x86/image/gen_image_efi.sh]: Should a failure affect the scripts `exit` value?  Like:

`|| { echo "Copying boot and EFI directory failed!" ; STATUS=1 }`

File: target/linux/x86/image/gen_image_efi.sh [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+dd if="$OUTPUT.kernel" of="$OUTPUT" bs=512 seek="$KERNELOFFSET" conv=notrunc || echo "Merging boot partition to disk image failed!"
+rm -f "$OUTPUT.kernel"
```
Comment (pprindeville) [inline target/linux/x86/image/gen_image_efi.sh]: add:

`exit $STATUS`

**Comment by alive4ever** [comment]
Sorry, I didn't do a rebase for a long time.

**Comment by pprindeville** [comment]
I have time to look at this and a machine to test it on...

**Comment by pprindeville** [comment]
> Sorry, I didn't do a rebase for a long time.

@alive4ever Could you?  And force push?  Then we could discuss some of the issues called out above... Thanks.

File: package/boot/grub2/Makefile [hunk 11]
```diff
@@ -51,44 +51,174 @@ define Package/grub2-editenv/description
 	Edit grub2 environment files.
 endef
 
+ifneq ($(CONFIG_GRUB_EFI),)
+GRUB_FONT:=unifont
+GRUB_FONT_VERSION:=9.0.06
+GRUB_FONT_SOURCE:=$(GRUB_FONT)-$(GRUB_FONT_VERSION).bdf.gz
+GRUB_FONT_SOURCE_URL:=@GNU/unifont
+GRUB_FONT_SOURCE_HASH:=4246c4773ed70f78a7e27ff1118fd257a280d1102200265ad5d58bb2011195ef
+
```
Comment (pprindeville) [inline package/boot/grub2/Makefile]: Can you also add a dependency on `CONFIG_KERNEL_FB_SIMPLE`?

**Comment by pprindeville** [comment]
@sduponch Are you still interested in getting something working?  I definitely need to be able to boot some x86_64 servers with NVMe to run LEDE, but they don't seem to like doing a BIOS boot with an MBR, etc.

**Comment by blogic** [comment]
With the remerge in progress, all PRs on the lede-project organisation will be closed. Please help getting this merged or rebase/post it on the openwrt project page (https://github.com/openwrt/openwrt/pulls). All remaining PRs will be closed in 30 days.

File: package/boot/grub2/Makefile [hunk 12]
```diff
@@ -51,44 +51,174 @@ define Package/grub2-editenv/description
...
+	$(call Host/Compile/bios,install,$(HOST_BUILD_PREFIX))
+endef
+
+define Host/Install/efi
+	$(INSTALL_DIR) $(STAGING_DIR_HOST)/share/fonts/$(GRUB_FONT)/
+	$(INSTALL_DATA) $(HOST_BUILD_DIR)/$(GRUB_FONT).bdf $(STAGING_DIR_HOST)/share/fonts/$(GRUB_FONT)/
+	$(call Host/Compile/efi,install,$(HOST_BUILD_PREFIX))
+endef
+
+ifneq ($(CONFIG_GRUB_EFI),)
```
[the remark this answers is not in the document]
Comment (alive4ever) [reply package/boot/grub2/Makefile]: Without conditional, efi images are always generated for x86_64. Is that ok?

**Comment by alive4ever** [comment]
Waiting for #1609 for gdisk required to generate gpt based image.

**Pull request closed without merging**